core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -114,6 +120,20 @@
     private final AtomicInteger throttleRequestCount = new AtomicInteger();
     private final EngineConfig.OpenMode openMode;
     private final AtomicBoolean allowCommits = new AtomicBoolean(true);
+    /**
+     * Refresh listeners. While they are not stored in sorted order they are processed as though they are.
+     */
+    private final LinkedTransferQueue<RefreshListener> refreshListeners = new LinkedTransferQueue<>();
```

Comment:
I had it there at first but I didn't like it there. It felt weird to have to tiptoe around ShadowEngine which doesn't really support this behavior.